test/hotspot/jtreg/runtime/Monitor/libMonitorWithDeadObjectTest.c

```diff
@@ -0,0 +1,172 @@
+  check(env, getThreadMXBeanMethod, "No getThreadMXBean method");
+
+  jobject threadBean = (*env)->CallStaticObjectMethod(env, ManagementFactoryClass, getThreadMXBeanMethod);
+  check(env, threadBean, "Calling getThreadMXBean()");
+
+  jclass ThreadMXBeanClass = (*env)->FindClass(env, "java/lang/management/ThreadMXBean");
+  check(env, ThreadMXBeanClass, "No ThreadMXBean class");
+
+  jmethodID dumpAllThreadsMethod = (*env)->GetMethodID(env, ThreadMXBeanClass, "dumpAllThreads", "(ZZ)[Ljava/lang/management/ThreadInfo;");
+  check(env, dumpAllThreadsMethod, "No dumpAllThreads method");
+
+  // The 'lockedMonitors == true' is what causes the monitor with a dead object to be examined.
+  jobject array = (*env)->CallObjectMethod(env, threadBean, dumpAllThreadsMethod, JNI_TRUE /* lockedMonitors */, JNI_FALSE /* lockedSynchronizers*/);
+  check(env, array, "Calling dumpAllThreads(true, false)");
+}
+
+static void create_monitor_with_dead_object(JNIEnv* env) {
+  jobject obj = create_object(env);
+
+  if ((*env)->MonitorEnter(env, obj) != 0) die("MonitorEnter");
+
+  // Drop the last strong reference to the object associated with the monitor.
+  // The monitor only keeps a weak reference to the object.
+  (*env)->DeleteLocalRef(env, obj);
+
+  // Let the GC clear the weak reference to the object.
+  system_gc(env);
+}
+
+static void* create_monitor_with_dead_object_in_thread() {
+  JNIEnv* env;
+  int res = (*jvm)->AttachCurrentThread(jvm, (void**)&env, NULL);
+  if (res != JNI_OK) die("AttachCurrentThread");
+
+  // Make the correct incantation to create a monitor with a dead object.
+  create_monitor_with_dead_object(env);
+
+  // DetachCurrenThread will try to unlock held monitors. This has been a
+  // source of at least two bugs:
+  // - When the object reference in the monitor was cleared, the monitor
+  //   iterator code would skip it, preventing it from being unlocked when
+  //   the owner thread detached, leaving it lingering in the system.
+  // - When the monitor iterator API was rewritten the code was changed to
+  //   assert that we didn't have "owned" monitors with dead objects. This
+  //   test provokes that situation and that asserts.
+  if ((*jvm)->DetachCurrentThread(jvm) != JNI_OK) die("DetachCurrentThread");
+
+  return NULL;
+}
+
+static void* create_monitor_with_dead_object_and_dump_threads_in_thread() {
+  JNIEnv* env;
+  int res = (*jvm)->AttachCurrentThread(jvm, (void**)&env, NULL);
+  if (res != JNI_OK) die("AttachCurrentThread");
+
+  // Make the correct incantation to create a monitor with a dead object.
+  create_monitor_with_dead_object(env);
+
+  // Perform a thread dump that checks for all thread's monitors.
+  // That code didn't expect the monitor iterators to return monitors
+  // with dead objects and therefore asserted/crashed.
+  thread_dump_with_locked_monitors(env);
+
+  if ((*jvm)->DetachCurrentThread(jvm) != JNI_OK) die("DetachCurrentThread");
+
+  return NULL;
```

Comment:
Why is this function return type "void*" when it only returns NULL?

Reply:
See earlier comment.